metals/src/main/scala/scala/meta/internal/metals/ImplementationProvider.scala

```diff
@@ -0,0 +1,121 @@
+package scala.meta.internal.metals
+import org.eclipse.lsp4j.Location
+import org.eclipse.lsp4j.TextDocumentPositionParams
+import scala.meta.internal.mtags.Semanticdbs
+import scala.meta.internal.metals.MetalsEnrichments._
+import scala.meta.io.AbsolutePath
+import java.nio.file.Paths
+import scala.meta.internal.semanticdb.TextDocuments
+import scala.meta.internal.semanticdb.SymbolOccurrence
+import scala.meta.internal.semanticdb.ClassSignature
+import scala.meta.internal.semanticdb.TypeRef
+import scala.meta.internal.semanticdb.Signature
+import scala.meta.internal.semanticdb.TextDocument
+import java.util.concurrent.ConcurrentHashMap
+import java.nio.file.Path
+
+final class ImplementationProvider(
+    semanticdbs: Semanticdbs,
+    workspace: AbsolutePath,
+    buffer: Buffers,
+    definitionProvider: DefinitionProvider
+) {
+  private val implementationsInPath =
+    new ConcurrentHashMap[Path, Map[String, Set[ClassLocation]]]
+
+  def implementations(params: TextDocumentPositionParams): List[Location] = {
+    val source = params.getTextDocument.getUri.toAbsolutePath
+
+    def findSemanticdb(fileSource: AbsolutePath) =
+      semanticdbs
+        .textDocument(fileSource)
+        .documentIncludingStale
+        .toList
+
+    for {
+      currentDoc <- findSemanticdb(source)
+      positionOccurrence = definitionProvider.positionOccurrence(
+        source,
+        params,
+        currentDoc
+      )
+      occ <- positionOccurrence.occurrence.toList
+      (file, locations) <- findImplementation(occ.symbol).groupBy(_.file)
+      fileSource = AbsolutePath(file)
+      doc <- findSemanticdb(fileSource)
```

Comment:
it is only used in a false branch of the next `if` statement. Maybe those could be merged?

Reply:
Actually, this is an optimization that most likely will not be needed. I will remove the `if else`, it clouds the logic somewhat